waspc/packages/deploy/src/common/waspBuild.ts

```diff
@@ -0,0 +1,32 @@
+import { type ProcessOutput } from "zx";
+import { WaspCliExe, WaspProjectDir } from "./brandedTypes.js";
+import { waspSays } from "./terminal.js";
+import { createCommandWithCwd } from "./zx.js";
+
+export const ensureWaspProjectIsBuilt = createEnsureWaspProjectIsBuilt();
+
+function createEnsureWaspProjectIsBuilt() {
+  let waspBuildResult: Promise<ProcessOutput> | undefined = undefined;
+
+  async function ensureWaspProjectIsBuilt({
+    waspProjectDir,
+    waspExe,
+  }: {
+    waspProjectDir: WaspProjectDir;
+    waspExe: WaspCliExe;
+  }): Promise<ProcessOutput> {
+    // We want to build the Wasp project only once per CLI invocation.
+    // Sometimes `ensureWaspProjectIsBuilt` is called multiple times
+    // (e.g. in `setup` command, we call it and then again when deploying).
+    if (waspBuildResult !== undefined) {
+      return waspBuildResult;
+    }
+
+    waspSays("Building your Wasp app...");
+    const waspCli = createCommandWithCwd(waspExe, waspProjectDir);
+    waspBuildResult = waspCli(["build"]);
+    return waspBuildResult;
+  }
+
+  return ensureWaspProjectIsBuilt;
+}
```

Comment:
Ok, this works! It's a bit weird following the logic, I'd do it like this, nicer separation between caching and logic:

```suggestion
function createEnsureWaspProjectIsBuilt() {
  async function ensureWaspProjectIsBuilt({
    waspProjectDir,
    waspExe,
  }: {
    waspProjectDir: WaspProjectDir;
    waspExe: WaspCliExe;
  }): Promise<void> {
    waspSays("Building your Wasp app...");
    const waspCli = createCommandWithCwd(waspExe, waspProjectDir);
    await waspCli(["build"]);
  }

  let cached: ReturnType<typeof ensureWaspProjectIsBuilt> | undefined = undefined

  // We want to build the Wasp project only once per CLI invocation.
  // Sometimes `ensureWaspProjectIsBuilt` is called multiple times
  // (e.g. in `setup` command, we call it and then again when deploying).
  return (...args: Parameters<typeof ensureWaspProjectIsBuilt>) => {
    cached ??= ensureWaspProjectIsBuilt(...args)
    return cached
  };
}
```

Not blocking on this as it's a style preference and we're short on time.